### DISCUSSION TITLE
tquic多路径能否为某一条流指定路径呢？如果不能，我应该修改哪一部分代码？

### DISCUSSION THREAD

**Discussion by wdj55556** [Tencent/tquic#289] [Q&A (问题和回答)]
我试图使用tquic来实现我的多路径算法，我希望把算法实现在客户端，为不同流指定路径，服务器只需要在请求的路径上回复。目前tquic好像不能为流指定路径，只能在发送数据包的时候选择路径。能给我一些修改tquic的建议吗？

**Comment by glime-ai** [comment]
根据现有的TQUIC文档和讨论，TQUIC目前确实只支持在发送数据包时选择路径，而不支持为不同的流指定路径。要实现你所描述的功能，可能需要对TQUIC进行一些修改。以下是一些建议：

1. **修改路径选择逻辑**：
   - 你可以在客户端实现一个自定义的路径选择逻辑，根据流的不同来选择不同的路径。具体来说，可以在发送数据包之前，根据流的ID或其他标识来决定使用哪条路径。

2. **扩展API**：
   - 你可能需要扩展TQUIC的API，以便在创建流时能够指定路径。这样可以在流创建时就确定路径，而不是在发送数据包时再选择。

3. **修改on_packet_send回调**：
   - 根据[讨论](https://github.com/Tencent/tquic/discussions/212)，`on_packet_send`回调函数负责最终将TQUIC生成的报文发送出去。你可以在这个回调中实现自定义的路径选择逻辑。

4. **参考现有的多路径实现**：
   - 你可以参考TQUIC现有的多路径实现，了解其工作原理，并在此基础上进行扩展。例如，TQUIC支持的多路径传输可以通过`enable_multipath`配置函数启用。

5. **文档和示例代码**：
   - 建议参考TQUIC的[官方文档](https://tquic.net/zh/docs/category/api-reference)和示例代码，了解其API和使用方法。你可以在这些资源中找到如何配置和使用多路径传输的详细信息。

如果你有更多具体的问题或需要进一步的帮助，建议在TQUIC的[GitHub讨论区](https://github.com/Tencent/tquic/discussions)提问，社区中的开发者可能会提供更多有价值的建议。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Comment by iyangsj** [comment]
tquic目前内置的多路径调度算法，不支持所述行为。

你可以自行进行定制开发，一些关键的修改包括：
1. 支持客户端通过[TransportParam](https://github.com/Tencent/tquic/blob/develop/src/trans_param.rs)协商及指定服务端所使用的多路径调度算法

2. 服务端增加的新调度算法，实现所述的发送策略。
需要注意的是，为了实现所述算法，可能需要扩展[多路径调度接口](https://github.com/Tencent/tquic/blob/develop/src/multipath_scheduler/multipath_scheduler.rs#L36-L60)。例如，增加on_recv接口，以便记录接收的Stream帧所使用的路径。
此外，需要修改报文发送行为。一种做法是由多路径调度器在 `on_select()`接口实现中，将流待发送的数据，放入对应路径的[发送缓冲队列](https://github.com/Tencent/tquic/blob/develop/src/connection/space.rs#L112)中。
